File: 프로그래머스/2/172927. 광물 캐기/광물 캐기.cpp
#include <string>
#include <vector>
#include <algorithm>
#include <iostream>

using namespace std;

int solution(vector<int> picks, vector<string> minerals) {
    int answer = 0;
    
    vector<array<int, 3>> score;
    array<int, 3> val = {0, 0, 0};
    int total = picks[0] + picks[1] + picks[2];
    int maxM = min((int)minerals.size(), total * 5);

    for (int i = 0; i < maxM + 1; i++) {
        if ((i % 5 == 0 && i != 0) || i == maxM) {
            
            score.push_back(val);
            val = {0, 0, 0};

            if (i == maxM) break;
        }
        if (minerals[i] == "diamond") val[0] += 1;
        if (minerals[i] == "iron") val[0] += 1;
        if (minerals[i] == "stone") val[0] += 1;
        
        if (minerals[i] == "diamond") val[1] += 5;
        if (minerals[i] == "iron") val[1] += 1;
        if (minerals[i] == "stone") val[1] += 1;
        
        if (minerals[i] == "diamond") val[2] += 25;
        if (minerals[i] == "iron") val[2] += 5;
        if (minerals[i] == "stone") val[2] += 1;
    }
    
    sort(score.begin(), score.end(), [&] (const auto &a, const auto &b) {
        return a[2] > b[2];
    });
    
    int select = 0;
    for (int i = 0; i < score.size(); i++) {
        while (picks[select] <= 0) {
            select++;
        }
        if (select >= 3) break;
        answer += score[i][select];
        picks[select]--;
    }
    return answer;
}

File: 프로그래머스/2/172927. 광물 캐기/광물 캐기.java
class Solution {
    int answer = Integer.MAX_VALUE;
    public int solution(int[] picks, String[] minerals) {
        dfs(0, 0, picks, minerals);
        return answer;
    }
    
    private void dfs(int index, int value, int[] picks, String[] minerals) {
        if (index == minerals.length || picks[0] + picks[1] + picks[2] == 0) {
            answer = Math.min(answer, value);
            return;
        }
        
        for (int i = 0; i < 3; i++) {
            if (picks[i] == 0) continue;
            int val = 0;
            int j;
            for (j = 0; j < 5 && j + index < minerals.length; j++) {
                val += cal(i, minerals[index + j]);
            }
            picks[i] -= 1;
            dfs(index + j, value + val, picks, minerals);
            picks[i] += 1;
        }
    }
    
    private int cal(int tool, String mineral) {
        if (tool == 0) return 1;
        if (tool == 1) {
            if (mineral.equals("diamond")) return 5;
            else return 1;
        }
        if (tool == 2) {
            if (mineral.equals("diamond")) return 25;
            else if (mineral.equals("iron")) return 5;
            else return 1;
        }
        return 0;
    }
}